### DISCUSSION TITLE
Extending QROptions

### DISCUSSION THREAD

**Discussion by nodescan-admin** [chillerlan/js-qrcode#3] [General]
Hi,

is it possible with this Javascript implementation to extend the QROptions as can be done with the PHP version? I would like to make circular QR codes with an svg logo in the center, like these examples.

https://github.com/chillerlan/php-qrcode/blob/main/examples/svgWithLogo.php
https://github.com/chillerlan/php-qrcode/blob/main/examples/svgRoundQuietzone.php

If it is possible but requires some custom javascript, I  would appreciate a brief example/starting point. If it's not possible I will look for other ways.

Thanks.

**Comment by codemasher** [comment]
Hi, yes the options class is supposed to work similar to the PHP version and can be extended (with some limitations).

https://github.com/chillerlan/js-qrcode/blob/2d66a5759533f64903b8ff6018178f8173c45fca/src/QROptions.js#L359-L388

**Reply by nodescan-admin** [reply]
Thank you for your reply and changing it to a Discussion. I did wonder if that might be better. I will try.

**Reply by codemasher** [reply]
I haven't yet updated the library (it's supposed to be similar to v5 of the PHP version), so some internals might not yet work as intended - use at your own risk :)

**Reply by nodescan-admin** [reply]
Hi,

I worked on translating the (https://github.com/chillerlan/php-qrcode/blob/main/examples/svgRoundQuietzone.php]  example to JavasScript. It is almost working with a few (hopefully minor) issues:

1) The circle is not completely filled in. It's a diamond shape...??

<img width="1339" alt="Screenshot 2024-03-04 at 20 07 31" src="https://github.com/chillerlan/js-qrcode/assets/91238962/2b015730-fd62-4327-9613-1542e927beb7">

2) The QR code is not diplayed in the browser but when I copy the SVG data from the console.log output and paste it into an online SVG viewer, it displays properly (apart from issue 1). 

3) 'dotColors' is logged as null even though it is the options... (Maybe related to 2 but I suspect not)  

If you have time could you please take a look? (Sorry the code formatting was messed up so I attached it a text file.)

[RoundQuietzoneSVGoutput.txt](https://github.com/chillerlan/js-qrcode/files/14480652/RoundQuietzoneSVGoutput.txt)

If you don't have time I will keep plugging away....
Thanks

**Reply by codemasher** [reply]
Hey, it would be super helpful if you'd post a self-contained example based on the HTML ones, so that I don't have to debug to get it running at all.
However, at a first glance, the IDE reported an error in the `checkIfInsideCircle()` method, that might be the issue:

```js
  if(Math.pow(($dx, 2) + Math.pow($dy, 2)) <= Math.pow($radius, 2)){
    return true;
  }  
```

It is supposed to be:
```js
  if((Math.pow($dx, 2) + Math.pow($dy, 2)) <= Math.pow($radius, 2)){
    return true;
  }  
```

[In PHP is just braced](https://github.com/chillerlan/php-qrcode/blob/900c8aea4f233cd8f3d3047de8978b987aa0a317/examples/svgRoundQuietzone.php#L135) to avoid ambiguity (according to my PHPCS settings).

**Comment by nodescan-admin** [comment]
Thank you. That was it on the main issue.  Very nice.  I will create a self-contained example and try to add the logo in the middle too.

<img width="1361" alt="RoundQuietzoneSVGoutput" src="https://github.com/chillerlan/js-qrcode/assets/91238962/72ab1dfe-f682-44f9-9f2f-30a157996bf3">

**Reply by codemasher** [reply]
Another thing i noticed is that the 1 module wide line is 4 modules away from the QR Code - it is supposed to be the quiet zone so that the symbol can be properly identified. Is this intentional?

**Reply by nodescan-admin** [reply]
No.  I see what you mean but thats's not intentional. I was trying to replicate your original php example. Do you have any idea which block/function to look at?

**Reply by codemasher** [reply]
It depends on the quiet zone size over here:

https://github.com/chillerlan/php-qrcode/blob/900c8aea4f233cd8f3d3047de8978b987aa0a317/examples/svgRoundQuietzone.php#L86-L108

and the call is over here:

https://github.com/chillerlan/php-qrcode/blob/900c8aea4f233cd8f3d3047de8978b987aa0a317/examples/svgRoundQuietzone.php#L38-L41

I guess there may be better ways to do this, but I don't have the brain juice for it currently.
